src/main/java/com/hivemq/cli/commands/hivemq/export/clients/ExportClientsCommand.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2019 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.hivemq.cli.commands.hivemq.export.clients;
+
+
+import com.google.common.base.Throwables;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.hivemq.cli.MqttCLIMain;
+import com.hivemq.cli.commands.hivemq.export.AbstractExportCommand;
+import com.hivemq.cli.rest.HiveMQRestService;
+import org.openapitools.client.ApiException;
+import org.openapitools.client.model.ClientDetails;
+import picocli.CommandLine;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+
+@CommandLine.Command(
+        name = "clients",
+        description = "Export HiveMQ client details",
+        sortOptions = false,
+        mixinStandardHelpOptions = true,
+        versionProvider = MqttCLIMain.CLIVersionProvider.class)
+public class ExportClientsCommand extends AbstractExportCommand implements Callable<Integer> {
+
+    private final static String DEFAULT_FILE_NAME = "hivemq_client_details";
+
+    @Inject
+    public ExportClientsCommand() { }
```

Comment:
Is there a particular reason for using the old concurrency APIs in this class?
I would welcome a refactoring to CompletableFutures and a more modern code base.

Reply:
Unfortunately, I didn't find a satisfying way to communicate errors from the 3 async tasks for error reporting issues because CompletableFuture may only throw unchecked exceptions.
Therefore, I ended up using the old api. But if you have any recommendations I will refactor the tasks to a more modern approach.